### ISSUE TITLE
Redundant station list on routes?

### ISSUE THREAD

**Issue by elliottwilliams** [propershark/shark#12]
A list of stations is included on routes (`route.stations`), and each station seems to be associated to its route (`route.associated_objects["Shark::Station"]`). Should one of these be dropped? And which list should I use?

See https://gist.github.com/elliottwilliams/a676a83fef3ecf9fa9443543f87ee590 for an example serialized route object.

**Comment by faultyserver** [comment]
I think we discussed this a little bit a few days ago, and from what I remember, we couldn't really come to a decision.

`route.stations` should definitely be kept, since it provides the station order (`associated_objects` is not an _ordered_ set). So, the question is really "should a route have associations to all of the stations it touches?".

Obviously, it's easiest to just leave it in (since it already works that way), but removing the duplication wouldn't be too difficult either.

Here's a pros/cons that I can see:

Pros:
- `associated_objects`, as a Set, is guaranteed to be a set of unique stations that a Route touches. This means that clients can avoid reading large (potentially 1000+ element) lists of stops that contain many duplicates (think small loops like Inner Loop or Ross Ade).
- Depending on the view, `associated_objects` can generically be used as "the set of topics to subscribe to" without any special logic for routes saying "for stations, use route.stations instead". This would also hit the first issue, where duplicate stations would be iterated whenever a route loops.
- Keeping the association simplifies the management of back-associations that stations have to the routes that pass through them. Since this information is updated through a route `update`/`activate` event, it is simplest to iterate the `associated_objects` during that event to dissociate removed stations and then add new ones (really, this is done by removing all associations and adding all of the new ones).

Cons:
- Data duplication is obviously useless. On large routes (that don't loop often), you could easily have 50+ station objects duplicated.
- The list doesn't provide much semantic information, Since order is not guaranteed, `associated_objects` can't be used to define the loop that a route follows (which, sans-duplicates, `route.stations` could), so the only information gained is the unique list of stations, which can be easily and quickly computed by the client.

Pro #3 is my primary rationale for keeping it. I also don't want to have special logic that masks attributes on events (though, that could be a separate middleware...hmmm), especially since this seems to be the only case for such logic right now.

My list isn't at all exhaustive. If you have any other pros/cons, feel free to add them. I honestly don't know much about how the client works/the nuances of using the information I'm publishing, so I wouldn't be surprised if there are things that I've missed.

**Comment by elliottwilliams** [comment]
Thanks. The biggest thing I was missing is how `route.stations` provides order and represents the entire trip, while `associated_objects` is just a set.

I think the way Proper should handle route stations is by treating associated stations as the canonical set of stations to be watched, and treating the list that `route.stations` provides as a "view" onto the station set, which imposes an order and cycles through the list of stations. (further explanation at propershark/proper#1)

With that in mind:

**Can the `name` attribute be passed in `associated_objects` and not `route.stations`?** That would prevent me from having to write some ugly/slow decoder code that would scan for a station's name in `route.stations` as it created parses each associated station. It would also cut down on the number of redundant bits sent since `routes.stations` could just be a list of ids, not (id, name) stub objects.

If not, perhaps Shark could send a `stationNames` dictionary with an `id:name` mapping? This would still make it easier to decode.

**Comment by faultyserver** [comment]
Previously, I was against modifying `associated_objects` to be anything other than lists of identifiers, but now that embedding is done in its own middleware, and is thus optional/configurable, I'm more okay with these modifications.

If I'm understanding you correctly, you want something like this:

``` js
"route": {
  "associated_objects": {
    "Shark::Station": [
      {
        "identifier": "station.BUS123",
        "name": "That station over there"
      },
      ...
    ]
  },
  ...,
  "stations": [
    "stations.BUS123",
    "stations.BUS456",
    ...
  ]
}
```

If so, that should be easily possible now.

**Comment by elliottwilliams** [comment]
That's right. Will ObjectEmbed be designed to be configurable per-event?

**Comment by faultyserver** [comment]
Ideally it would be, but the current Configuration class doesn't have a way of doing that.

I'm debating making a separate MiddlewareConfiguration that provides `on_event` hooks and other such conveniences, but I haven't started that yet.

**Comment by elliottwilliams** [comment]
Cool. It's fairly low-priority to me. For the time being, I will probably just use station identifiers

**Comment by faultyserver** [comment]
The merge of #18 makes this possible, and it is halfway complete.

All that's left is to _stop_ embedding objects in `route.stations`, leaving it as a list of identifiers.

**Comment by faultyserver** [comment]
This should be fixed with 82da624. This commit made it so that attributes that represent Objects are simply _included_ by default, rather than _embedded_. Embedding is an opt-in functionality for each attribute. See the commit message for more details.

**Comment by faultyserver** [comment]
Did a bit more testing, it looks like this is working properly now. Closing as completed.
